File: Dictionaries and Sets/1365. How Many Numbers Are Smaller Than the Current Number/1365. How Many Numbers Are Smaller Than the Current Number.py
class Solution:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        dictt = {}
        for i, n in enumerate(sorted(nums)):
            if n not in dictt:
                dictt[n] = i 
        
        ans = []
        for n in nums:
            ans.append(dictt[n])
        
        return ans

# Time: O(n log n)
# Space: O(n)

File: Dictionaries and Sets/1365. How Many Numbers Are Smaller Than the Current Number/1365. How Many Numbers Are Smaller Than the Current Number.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallerNumbersThanCurrent = function (nums) {
  const sortedNums = nums.slice().sort((a, b) => a - b);
  let numsLess = {};
  for (let i = 0; i < sortedNums.length; i++) {
    if (!(sortedNums[i] in numsLess)) numsLess[sortedNums[i]] = i;
  }

  let ans = [];
  for (let n of nums) ans.push(numsLess[n]);
  return ans;
};

// Time: O(n log n)
// Space: O(n)